desc/protoparse/ast/file.go

```diff
@@ -0,0 +1,230 @@
+package ast
+
+import "fmt"
+
+// FileNode is the root of the AST hierarchy.
+type FileNode struct {
+	compositeNode
+	Syntax   *SyntaxNode // nil if file has no syntax declaration
+
+	// Any package declarations in the file. Note that a valid file
+	// will have only zero or one such declaration.
+	Package  []*PackageNode
+	Imports  []*ImportNode
+	Options  []*OptionNode
+	Messages []*MessageNode
+	Enums    []*EnumNode
+	Extends  []*ExtendNode
+	Services []*ServiceNode
+
+	// All of the above elements in their original order as encountered
+	// in the source file.
+	AllDecls    []FileElement
```

Comment:
Why is this needed?

Reply:
So you can iterate the declaration in their original order. The fields above just group them by concrete type. A code formatter, for example, would want to use this, not the fields above, so that it can emit declarations in original source order (possibly interleaved across concrete types).

Reply:
I am wondering if this should be an either/or thing, as opposed to exposing both - perhaps the groupings by type can be exposed as static functions on each type, as opposed to being fields?

Reply:
Can you elaborate on your concern about exposing both as fields? Is the worry about mutating the AST, and being concerned with _where_ to make changes? I would think for changing the AST, you would use the `New*` functions to create new nodes and then rebuild all of the ancestors -- treating the existing nodes more or less as if they are immutable.

I dislike that Go doesn't provide any help for immutable value types because that's really what we want here. But the overhead of actually making them immutable is high -- lots of boiler-plate, and proliferation of types to model containers thanks to lack of generics (just like in the protobuf API v2 stuff for descriptors).

Reply:
Nice, this also leaves room for an opinionated formatter to iterate over each _type_ and format according to its specification. In that case, we wouldn't use  `AllDecls`, but I see why they're captured here.

Reply:
I ended up removing the grouped slices and leave only a `Decls []FileElement` slice. This change already has to allocate and compute many more slices for tracking the full tree (every non-leaf node has a slice of its children), so having to also allocate and compute all of these other slices seemed like too much overhead.

I'm still wondering if the impl of `compositeNode` needs to change so that it doesn't require so much slice creation. That could mean a lot of gross boiler-plate in each concrete non-leaf node type though...